### PR TITLE
Add cider.tasks/add-middleware boot task

### PR DESCRIPTION
**Pull request by arichiardi**
This adds the task for https://github.com/clojure-emacs/cider/pull/1751

Before submitting a PR make sure the following things have been done:
- [X] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [X] You've added tests to cover your change(s) (see `cider`)
- [X] All tests are passing (_Not really, but I don't think it is on my side_)
- [X] The new code is not generating reflection warnings

In order to try the command:

```
boot -v -d org.clojure/tools.nrepl:0.2.12 -d cider/cider-nrepl:0.15.0-SNAPSHOT -i "(require 'cider.tasks)" cider.tasks/add-middleware -m cider.nrepl/cider-middleware
```

And check the final few lines of the debug log:

```
Boot's default middleware: []
After cider's injection: [cider.nrepl/cider-middleware]
```

### CODE REVIEW

**Comment by Malabarba** [comment]
Thanks again. The code looks valid. I'll checkout the branch and test locally tonight.

**Comment by arichiardi** [comment]
No problem, a bit sloppy this one, not a lot of time but I wanted to see it in :+1: Thanks for the review.

**Comment by arichiardi** [comment]
@Malabarba I am trying to install this locally with:
`lein with-profile plugin.mranderson/config install`

but I have an empty jar, is the above the right command?

File: src/cider/tasks.clj [hunk 1]
```diff
@@ -0,0 +1,17 @@
+(ns cider.tasks
+  (:require [boot.core :refer [deftask]]
+            [boot.repl :as repl]
+            [boot.util :as util]))
+
+(deftask add-middleware
+  "CIDER middleware injection task
```
Comment (bbatsov) [inline src/cider/tasks.clj]: Is this the format of a boot task docstring? It looks a bit odd to me.

File: src/cider/tasks.clj [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  This task allows to inject middleware in boot.repl/*default-middleware*.
```
Comment (bbatsov) [inline src/cider/tasks.clj]: Wrap `boot.repl/*default-middleware*` in backticks.

File: src/cider/tasks.clj [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  Just pass them as -m|-middleware. The input is a string but will be converted
+  to symbol."
+  [m middleware MW #{str} "String of the middleware to inject"]
```
Comment (bbatsov) [inline src/cider/tasks.clj]: String -> Name

Comment (bbatsov) [inline src/cider/tasks.clj]: Btw, I'm a bit confused about this as well. What's `MW`?

File: src/cider/tasks.clj [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  (if-let [default-middleware (resolve 'boot.repl/*default-middleware*)]
+    (do (swap! @default-middleware concat (map symbol middleware))
+        (util/dbug "Injected middleware: %s\n" (vec @@default-middleware)))
```
Comment (bbatsov) [inline src/cider/tasks.clj]: Won't this show both default middleware + the injected middleware?

**Comment by bbatsov** [comment]
Apart from my small remarks - squash your commits and rebase on top of master.

File: src/cider/tasks.clj (hunk 3 above)
Comment (arichiardi) [inline src/cider/tasks.clj]: Boot tasks usually take a symbol there to identify the argument: https://github.com/boot-clj/boot/wiki/Task-Options-DSL

File: src/cider/tasks.clj (hunk 4 above)
Comment (arichiardi) [inline src/cider/tasks.clj]: Yes, default is what boot injected and plus we show what we injected, maybe we can actually have two different log calls here.

File: src/cider/tasks.clj (hunk 2 above)
Comment (arichiardi) [inline src/cider/tasks.clj]: K

File: src/cider/tasks.clj (hunk 1 above)
Comment (arichiardi) [inline src/cider/tasks.clj]: Yes it is, see DSL below

**Comment by arichiardi** [comment]
I am going to give it a final go with the new changes (to both this and `cider's` split) and report here if all is good.

**Comment by arichiardi** [comment]
Ok tested with a (cljs) boot project and (clj) lein project. Good to go!

**Comment by bbatsov** [comment]
👍

**Pull request merged**